\documentclass{article}
\usepackage{enumitem}
\usepackage{fixltx2e}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{listings}
\usepackage{color}
\usepackage[tbtags]{amsmath}
\usepackage{textcomp}
\usepackage{subcaption}
\usepackage[top=25truemm,bottom=30truemm,left=25truemm,right=25truemm]{geometry}



\setlength\intextsep{5pt}
\setlength\textfloatsep{5pt}


\def\vector#1{\mbox{\boldmath $#1$}}
\begin{document}

\title{Inter-domain Relationships in The Structure of Complex Network}
\author{Kansuke Ikehara}
\maketitle

\begin{abstract}
The structure of complex networks has been of interest in many scientific and engineering disciplines over decades. A number of scientists have reported findings that indicated the characteristics of network structure that are claimed to be universal, such as \textit{small-world} and \textit{small-world} networks. Domain specific characteristics, for example specific profiles of network motifs and high-triangle density in social networks, have also been reported previously. However, there is no comprehensive study of inter-domain relationships in the structure of various complex networks. In this thesis, we study over 1200 networks of diverse domains ranging from ecological food-web to online social network using machine learning techniques, such as random forest and confusion matrix and show the relationships among network domains in terms of network structure. Our result indicates that there are some pairs of network domains (categories) that are inherently hard to distinguish based purely on network structure. We have found that these domains tend to have similar underlying mechanisms which generate a network or drive the growth of a network or processes which happen on a network itself.

\end{abstract}
\tableofcontents


\section{Introduction}
	\subsection{Complex Networks}
	Almost every scientific and engineering discipline deals with data that come from experimental observations or simulations running on a computer. Traditionally, data have been numerals, which may represent temperature, velocity, or voltage, and methodologies that analyze these numerals have been established for over hundreds of years. 
	
	\textit{Graph} or \textit{network}, which is essentially structured data as opposed to simple numeric values, has been used as a new way to approach real-world problems in last decades. Social science, for instance, has witnessed the power of network analysis with the recent emergence of online social network services such as Facebook, Twitter, LinkedIn, etc.  Social scientists have been solving previously intractable questions in the field with the large amount of  data and advanced graph algorithms with high-performance computers. Biological sciences use networks as a tool to dissect biological, chemical and even ecological processes in order to gain insight into how neurons in our brain interact with each other, how nutrients get broken down into pieces and get re-assembled in another molecular form, and how animals eat or be eaten by other animals. Engineering systems such as the Internet, power grids, water distribution networks, transportation networks, etc. have also been investigated with using network analysis tools for constructing efficient and robust systems. 
	
	
	\subsection{Structural Universality of Networks}
	
	One of the main research themes in Network Science, an interdisciplinary field which studies properties of networks ranging from biological and social networks to engineering system networks, is to find hidden universal patterns in the networks of various categories. 
	Watts and Strogatz proposed a random network model that, depending on the parameter setting, produces a network having properties as follows: 1. high density of triangles, implying that if three nodes are connected, it is likely that those three nodes actually compose a triangle; 2. low average distance between a pair of nodes, which indicates that from any node it is just a few steps, in average, to reach any other node in the network \cite{watts1998cds}. Networks having these attributes are often called "small-world" networks and many kinds of networks have been reported as having the small world property \cite{Network_Small-World-Ness}.

	
	
	 Another property that is claimed to be universal by some researchers, yet that is still controversial, is \textit{power-law} degree distribution which is sometimes referred to as \textit{scale-free}. In a scale-free network, one observes a number of nodes with few connections while there are few nodes with many connections to others. This very skewed distribution of degree, connections attached to a node, is claimed to be found among diverse sets of networks \cite{Barabasi99emergenceScaling}. However, claims that a network follows the power-law degree distribution often lacks a valid statistical treatment in order to show it is indeed following the distribution \cite{Clauset:PowerLaw}


	\subsection{Structural Diversity of Networks}
	Contrary to prevalent small-world and scale-free properties, each category of networks exhibits unique structural profiles of its own. Social networks have been investigated by many researchers and found out that they exhibit unique structural characteristics: high density of triangle (clustering coefficient) and positive degree assortativity \cite{WhySocialNetworks, Mislove:2007:OnlineSocial}. High triangle density indicates the high probability of one's friends being friends each other and the positive  degree assortativity shows a tendency that high-degree nodes (nodes having many connections) connect to other high-degree nodes while low-degree nodes connect to other low-degree nodes.
	
	Roth \textit{et al.} identified a unique structure of subway networks, that is called core-branch structure in the literature \cite{Train}. Many real world subway networks have a ring-shape connections of stations and dense connections therein, referred to as the "core" of the network. In the core, stations are relatively densely connected to each other, enabling residents of a city to move around quickly. From the ring of the core, branches radiate outward connecting stations far from the center of the city. 
	
	Milo \textit{et al.} introduced a concept called \textit{network motifs} that are essentially patterns of frequent sub-graph (motif) in a particular network compared to its randomized network\cite{Milo_motif}. They have shown each category of network, such as gene regulation (transcription), food webs, electronic circuits, etc., has a distinct pattern of network motifs. Furthermore, Milo \textit{et al.} revealed the existence of superfamilies of networks that  are groups of network categories having the highly convergent motif profiles \cite{Milo_SuperFamily}. Onnela \textit{et al.} constructed a taxonomy of 746 networks  of diverse categories using "Mesoscopic Response Functions" (MRFS) that essentially describes the changes of specific values related to the community structure of a network with respect to a parameter $\xi \in [0,1]$\cite{Onnela_Taxonomy}. Each network has its own MRFS and they calculated distance between networks which is defined as the area of difference between two networks' MRFS. They have successfully identified clusters of similar networks in a dendrogram constructed based on MRFS distance.
	
	\subsection{Searching The Underlying Principles}
Network is merely a simplified mathematical object and yet, one could gain some insight into a phenomenon and/or process of a studied field just from network structure. 
If there is a diverse set of networks of various kinds and scales, what can we say about inter-domain relationships between network categories? The question is threefold: 
\begin{enumerate}
	\item \textit{What aspects of network structure make a specific category of network different from others?}
	\item \textit{Are there any sets of network categories that are inherently indistinguishable from each other based on network structure?} 
	\item \textit{If two networks of different categories are indistinguishable by network structure, are their mechanisms of underlying processes the same? And vice versa.}
\end{enumerate}

The first part of the question is, in some sense, an extension to studying social networks' structural uniqueness. As far as we know, no previous study has done extensive investigation in finding unique characteristics of specific kinds of networks compared to other kinds except social network.

The second part of the question asks if there is any structural similarity between different kinds of networks. We, however, use the word "indistinguishable" in stead of "similar" since we try to observe the commonality from a confusion matrix of a classifier, where a misclassified instance is considered to be indistinguishable from a class it's classified as because it is so similar to other instances of the wrong class the algorithm fails to label the instance correctly.


The last part of the question is essentially all about elucidating the meta-structure in domains or class labels. What if two very distinct categories (labels) of networks, say biological and technological ones, exhibit very similar network structure and a classifier misclassifies them many times? Is this because their underlying network generative processes, or their processes on networks themselves are the same? Answering this question helps us understand a driving mechanism for network structure.


The contribution of answering those questions comes in two ways: 1 (scientific point). it gives us a general conceptual framework upon which networks are studied across domains. Previous studies have only looked at a single category or multiple categories as one category, ignoring the relationships between categories. By studying networks across domains, one could find general theories of networks or test a hypothesis across domains in a more plausible way.  For instance, one could test validity of a network model across all of domains and find out which domains the network model can explain well; 2 (practical point). it gives us the knowledge base upon which various network-related algorithms could be constructed and tuned properly.  Many practical graph algorithms take no assumption in domain-specific network structure. It may be possible, however, to construct a new algorithm which runs faster and more efficiently on a specific kind of networks by taking into account of such domain-specific knowledge. For example, if there exists any unique network structural property of recommendation networks,  it may be applied to construct or fine tune a recommendation engine that utilizes knowledge of unique network structure. As the size of networks we obtain has grown in an unprecedented scale, domain-specific knowledge in network structure will be a key to analyze such large networks in a faster and more efficient way.

\subsection{Project Description}

In this thesis, we study 1200 (the number may change later) of various kinds of networks, ranging from ecological food-web to online social network to digital circuit. We extract network statistics as features of a network, construct a high-dimensional feature space and map each network onto the feature space. We then train a machine learning algorithm called \textit{random forest} with the training set of network data and let it classify the test networks based on their features. As the category distribution is skewed, meaning that some categories of networks have larger number of instances than other minority categories, we try several sampling strategies and show the effect of each methodology. We construct confusion matrices based upon classification results and proceed to analyze the misclassification with which we can answer research questions explained in the previous section. We then conclude with discussion based on several hypotheses and classification results.



\section{What Is A Network?}
	\subsection{Formal Definition}
	Formally, a network or graph is a mathematical object consisting of a set of edges (arcs) and nodes (vertices), which can be written as $G = (E,V)$. In many cases, an \textit{adjacency matrix} is used as a representation of a network, where each element of the matrix $A_{ij}$ takes a binary value, $1$ for presence of an edge between nodes $i$ and $j$, and $0$ if there is no edge between these nodes in an unweighted network, or real value if edges of the network are weighted. If the matrix is symmetric, namely $A_{ij} = A_{ji},  \forall i,j \in V$, the matrix represents an \textit{undirected} network in which edges do not have directionality at all. If the matrix is asymmetric, on the other hand, it's representing a \textit{directed} network, where edges have directionality. If the diagonal elements of a matrix $A$ are non-zero, the network has \textit{self-loops} which indicates that there are edges originating from and pointing to the exact same node. In many studies such self-loops are simply ignored for the sake of simplicity and out study is no exception as well.
	
	\subsection{Structural Features of A Network}
	There are many ways to characterize a network in a quantitative manner that use the network's structural features \cite{Newman:NetworksIntro,NetworkCharacterizationSurvey}. In the following sections we describe important structural features that are relevant in this study.
		
	 
	\subsubsection{Degree}
	A \textit{degree} of a node in a network is essentially the number of edges attached to the node. For a node $i$ in an unweighted network, it can be written mathematically as:
	
	\begin{equation}
 	 k_i = \sum_{j = 1}^n A_{ij}.
	\end{equation}
	
	In a directed network, there are two kinds of degree, namely \textit{in-degree} and \textit{out-degree}, each of which represents a degree of a node $i$ for incoming edges and outgoing edges respectively:
	
	\begin{equation}
 	 k_i^{in} = \sum_{j = 1}^n A_{ij},
	\end{equation}
	
	\begin{equation}
 	 k_i^{out} = \sum_{i = 1}^n A_{ij}.
	\end{equation}

	
	\subsubsection{Clustering Coefficient}
	Clustering coefficient, which describes a degree of \textit{transitivity} in a network, is one of the most widely used metrics in network analysis especially in the context of social network. Transitivity in the context of networks means that if nodes $a$ and $b$ are connected as well as nodes $b$ and $c$, then nodes $a$ and $c$ are connected. Mathematically, the definition of clustering coefficient is given by the following equation:
	\begin{equation}
	C = \frac{\text{number of closed paths of length two}}{\text{number of paths of length two}}.
	\end{equation}

	\subsubsection{Degree Assortativity}
	\textit{Assortativity} in a network indicates a tendency for nodes to be connected to other nodes with similar node attribute. In a social network, for instance, node attributes which could be the basis for assortativity include language, age, income, alma mater, etc. On the other hand, \textit{disassortativity} exhibits nodes' tendency to be connected to other nodes having different node attribute in the network. An example network of disassortatvity would be a social network of sexual-relationship among people. 
	\textit{Degree assortativity} is a form of assortativity in which nodes with similar degree value tend to be connected together. Therefore in a network exhibiting high degree assortativity, there is a core of highly connected nodes with high degree and periphery of nodes sparsely connected to other low-degree nodes.
	
	Here, we first derive a general \textit{assortativity coefficient} of an undirected, unweighted network based on scalar node attribute, excluding enumerative ones such as language, alma mater, etc., then we easily find \textit{degree assortativity coefficient}.
	
	First we calculate the covariance of $x_i$ and $x_j$ for the vertices at the ends of all edges, where $x_i$,$x_j$ are nodes $i$ and $j$'s attributes.
	The mean $\mu$ of the value of $x_i$ at the end of one edge is defined as:
	\begin{equation}
	\mu = \frac{ \sum_{ij}A_{ij}x_i }{\sum_{ij}A_{ij}} = \frac{\sum_i k_i x_i}{\sum_i k_i} = \frac{1}{2m}\sum_i k_i x_i,
	\end{equation}
where $k_i$ is a degree of node $i$.

Then, the covariance of of $x_i$ and $x_j$ over all edges is derived as:
	\begin{equation}
	\begin{split}
	\text{cov}(x_i, x_j)  &= \frac
					{ \sum_{ij}A_{ij}(x_i - \mu)(x_j - \mu) }
					{\sum_{ij}A_{ij} } \\
					&= \frac{1}{2m}\sum_{ij}A_{ij} (x_i x_j - \mu x_i - \mu x_j +\mu^2) \\
					&= \frac{1}{2m}\sum_{ij}A_{ij}x_i x_j - \mu^2 \\
					&= \frac{1}{2m}\sum_{ij}A_{ij}x_i x_j - \frac{1}{(2m)^2}\sum_{ij}k_i k_j x_i x_j \\
					&= \frac{1}{2m}\sum_{ij}(A_{ij} - \frac{k_i k_j}{2m})x_i x_j.
	\end{split}
	\end{equation}
This covariance will be positive if both $x_i$ and $x_j$ have, in average, similar values, and will be negative if both $x_i$ and $x_j$ tend to change in a different direction. We normalized cov($x_i$,$x_j$) by another quantity which represents a perfect assortativity, so that it takes a value $r \in [-1 ,1]$. The perfect matching happens if $x_i = x_j$ for all edges, and cov($x_i$,$x_i$) becomes:
	\begin{equation}
	\begin{split}
	 \text{cov}(x_i, x_i) &=  \frac{1}{2m}\sum_{ij}(A_{ij} x_i^2 - \frac{k_i k_j}{2m} x_i x_j) \\
	 		             &=  \frac{1}{2m}\sum_{ij}(k_i \delta_{ij} - \frac{k_i k_j}{2m}) x_i x_j,
	\end{split}
	\end{equation}
	where $\delta_{ij}$ is Kronecker delta.
Thus, the normalized covariance becomes as follows:
	\begin{equation}
	\begin{split}
	 r = \frac{ \text{cov}(x_i, x_j)}{ \text{cov}(x_i, x_i)} = 
	 \cfrac{\sum_{ij}(A_{ij} - \cfrac{k_i k_j}{2m})x_i x_j}{\sum_{ij}(k_i \delta_{ij} - \cfrac{k_i k_j}{2m}) x_i x_j}.
	\end{split}
	\end{equation}

Degree assortativity coefficient can easily be obtained by substituting $x_i$ and $x_j$ with degrees of respective vertices, thus:
	\begin{equation}
	 r =  \cfrac{\sum_{ij}(A_{ij} - \cfrac{k_i k_j}{2m})x_i x_j}{\sum_{ij}(k_i \delta_{ij} - \cfrac{k_i k_j}{2m}) k_i k_j}.
	\end{equation}

	\subsubsection{Network Motifs} 
The idea of \textit{network motifs} was first introduced by Milo \textit{et al.} \cite{Milo_motif}. Network motif is a sub-graph of a network that appears statistically significant than in its randomized counterpart, usually random network having the same degree distribution. There are a number of studies using network motifs of directed networks, where edges have directions, especially in biological sciences \cite{Alon2007, MotifsInBrain}. We, however, only use "connected" network motifs of undirected networks, that produce fewer variations in motif kinds, but can be applied to any network regardless of edge directionality which is crucial for our study. Fig. \ref{motifs} shows the complete list of $k=4$ undirected connected motifs used in this study.

\begin{figure}[ht]
	\begin{center}
		\vspace{0.5cm}
		\includegraphics[clip,width=12cm,height = 3cm]{motifs.png}
		\vspace{0.5cm}
		\caption{Network motifs of $k =4$, where $k$ is the number of node in a connected sub-graph.}
		\label{motifs}
	\end{center}
\end{figure}

In order to quantify network motifs, we first count the occurrence of each sub-graph in the original network, then repeat the same process on the randomized networks that have the exact same degree distribution. Such random networks are often referred to as \textit{configuration model} and are widely used as a null model for calculating various statistics related to networks. After counting occurrences of sub-graphs in both original and multiple random networks, we proceed to calculate z-score for each sub-graph $i$ as follows:
	\begin{equation}
	Z_i = \cfrac{N_i^{\text{original}} - \langle N_i^{\text{random}} \rangle }{\sigma_i^{\text{random}}},
	\end{equation}

where $N_i^{\text{original}}$ is the number of occurrence of a sub-graph $i$ in the original network and $ \langle N_i^{\text{random}} \rangle$ and $\sigma_i^{\text{random}}$ are the average and the standard deviation of the number of occurrence of a sub-graph $i$ in an ensemble of random networks. It is usually convenient to normalize this z-score as some networks exhibits very large values due to the size of the networks. Such normalized z-score is called \textit{significance profile} and is defined as follows:
	\begin{equation}
	SP_i = \cfrac{Z_i}{\sum_j Z_j^2}.
	\end{equation}

\section{Data Sets}
	
	Our network data set has been accumulated with a lot of effort over years and hyper links to most of the data are available at \url{https://icon.colorado.edu}.
	Since data format of real world networks is not standardized, we proceeded to convert all the data into a single format called \textit{Graph Modeling Language} or GML. The format allows us to flexibly specify arbitrary node and edge attributes. However, 
	
	

\section{Methods}

	\subsection{Sampling Strategies}
		\subsubsection{Random Over Sampling}
		\subsubsection{Random Under Sampling}
		\subsubsection{SMOTE}
	
	\subsection{Classification}
		\subsubsection{Random Forest Classifier}
		\subsubsection{Confusion Matrix}
		\subsubsection{}
		
	\subsection{}

\section{Analyses}
\subsection{}

\bibliographystyle{ieeetr}
\bibliography{reference} 

\end{document}












 
 